File: typescript/src/providers/google.ts
import { createRequire } from 'node:module';
import { computeCost, LLMMessage, LLMProvider, LLMResponse, LLMUsage, ProviderOptions } from './base.js';

const _req = createRequire(import.meta.url);

export class GoogleProvider implements LLMProvider {
  _genai: any;
  private session: any;
  private circuitBreaker: any;
  private defaultModel: string;

  constructor(opts: ProviderOptions & { baseUrl?: string } = {}) {
    const key = opts.apiKey ?? process.env['GOOGLE_API_KEY'];
    if (!key) throw new Error('GOOGLE_API_KEY not set and apiKey not provided');

    let mod: any;
    try {
      mod = _req('@google/generative-ai');
    } catch {
      throw new Error('google-generativeai required: npm install @google/generative-ai');
    }

    const Ctor = mod.GoogleGenerativeAI ?? mod.default?.GoogleGenerativeAI ?? mod;
    this._genai = new Ctor(key);
    this.session = opts.session;
    this.circuitBreaker = opts.circuitBreaker;
    this.defaultModel = opts.defaultModel ?? 'gemini-1.5-flash';
  }

  async complete(
    messages: LLMMessage[],
    model?: string,
    opts: Record<string, unknown> = {},
  ): Promise<LLMResponse> {
    const modelId = model ?? this.defaultModel;

    if (this.circuitBreaker) {
      const trip = this.circuitBreaker.checkAndRecord(messages[0]?.content ?? '', 0, Date.now());
      if (trip !== null) {
        const { BudgetExceeded } = await import('../errors.js');
        throw new BudgetExceeded(`Circuit breaker tripped: ${String(trip)}`);
      }
    }

    // Separate system messages — Gemini prepends to first user turn
    let systemContent = '';
    const conversation: LLMMessage[] = [];
    for (const msg of messages) {
      if (msg.role === 'system') {
        systemContent = msg.content;
      } else {
        conversation.push(msg);
      }
    }

    if (conversation.length === 0) {
      throw new Error('At least one user or assistant message is required');
    }

    const history = conversation.slice(0, -1).map(msg => ({
      role: msg.role === 'assistant' ? 'model' : 'user',
      parts: [{ text: msg.content }],
    }));

    const last = conversation[conversation.length - 1];
    const userText = systemContent ? `${systemContent}\n\n${last.content}` : last.content;

    const genModel = this._genai.getGenerativeModel({ model: modelId });
    const chat = genModel.startChat({ history });
    const result = await chat.sendMessage(userText);
    const response = result.response;

    const inputTokens: number = response.usageMetadata?.promptTokenCount ?? 0;
    const outputTokens: number = response.usageMetadata?.candidatesTokenCount ?? 0;
    const costUsd = computeCost(modelId, inputTokens, outputTokens);

    if (this.session) {
      this.session.recordUsage(inputTokens, outputTokens, costUsd, modelId);
    }

    return {
      text: response.text(),
      model: modelId,
      usage: { inputTokens, outputTokens, costUsd } satisfies LLMUsage,
      raw: response,
    };
  }
}

File: python/src/truss/providers/google.py
from __future__ import annotations

import os
import time
from typing import Any, Optional

from truss.providers.base import LLMMessage, LLMResponse, LLMUsage, compute_cost


class GoogleProvider:
    def __init__(
        self,
        api_key: Optional[str] = None,
        session: Any = None,
        circuit_breaker: Any = None,
        default_model: str = "gemini-1.5-flash",
    ) -> None:
        try:
            import google.generativeai as _genai
        except ImportError:
            raise ImportError(
                "google-generativeai required: pip install truss-ai[google]"
            ) from None

        key = api_key or os.environ.get("GOOGLE_API_KEY")
        if not key:
            raise ValueError("GOOGLE_API_KEY not set and api_key not provided")

        _genai.configure(api_key=key)
        self._genai = _genai
        self._session = session
        self._circuit_breaker = circuit_breaker
        self._default_model = default_model

    def complete(
        self,
        messages: list[LLMMessage],
        model: Optional[str] = None,
        **opts: Any,
    ) -> LLMResponse:
        from truss.errors import BudgetExceeded

        model_id = model or self._default_model

        if self._circuit_breaker:
            prompt = messages[0].content if messages else ""
            trip = self._circuit_breaker.check_and_record(prompt, 0.0, int(time.time() * 1000))
            if trip is not None:
                raise BudgetExceeded(f"Circuit breaker tripped: {trip.value}")

        # Separate system message — Gemini prepends it to the first user turn
        system_content = ""
        conversation: list[LLMMessage] = []
        for msg in messages:
            if msg.role == "system":
                system_content = msg.content
            else:
                conversation.append(msg)

        if not conversation:
            raise ValueError("At least one user or assistant message is required")

        # Build Gemini history (all turns except the last user message)
        history = []
        for msg in conversation[:-1]:
            gemini_role = "model" if msg.role == "assistant" else "user"
            history.append({"role": gemini_role, "parts": [msg.content]})

        last = conversation[-1]
        user_text = f"{system_content}\n\n{last.content}".strip() if system_content else last.content

        genai_model = self._genai.GenerativeModel(model_id)
        chat = genai_model.start_chat(history=history)
        result = chat.send_message(user_text)
        response = result.response

        input_tokens = getattr(getattr(response, "usage_metadata", None), "prompt_token_count", 0) or 0
        output_tokens = getattr(getattr(response, "usage_metadata", None), "candidates_token_count", 0) or 0
        cost = compute_cost(model_id, input_tokens, output_tokens)

        if self._session is not None:
            self._session.record_usage(
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                cost_usd=cost,
                model=model_id,
            )

        return LLMResponse(
            text=response.text,
            model=model_id,
            usage=LLMUsage(input_tokens=input_tokens, output_tokens=output_tokens, cost_usd=cost),
            raw=response,
        )
